homeassistant/components/aquacell/coordinator.py

```diff
@@ -0,0 +1,84 @@
+"""Coordinator to update data from Aquacell API."""
+
+import asyncio
+from datetime import timedelta
+import logging
+
+from aioaquacell import (
+    AquacellApi,
+    AquacellApiException,
+    AuthenticationFailed,
+    Softener,
+)
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_EMAIL, CONF_PASSWORD
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryAuthFailed, ConfigEntryError
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import CONF_REFRESH_TOKEN, UPDATE_INTERVAL
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class AquacellCoordinator(DataUpdateCoordinator[list[Softener]]):
+    """My aquacell coordinator."""
+
+    config_entry: ConfigEntry
+
+    def __init__(self, hass: HomeAssistant, aquacell_api: AquacellApi) -> None:
+        """Initialize coordinator."""
+        super().__init__(
+            hass,
+            _LOGGER,
+            name="Aquacell Coordinator",
+            update_interval=timedelta(seconds=UPDATE_INTERVAL),
+        )
+
+        self.refresh_token = self.config_entry.data[CONF_REFRESH_TOKEN]
+        self.email = self.config_entry.data[CONF_EMAIL]
+        self.password = self.config_entry.data[CONF_PASSWORD]
+        self.aquacell_api = aquacell_api
+
+    async def _async_update_data(self) -> list[Softener]:
```

Comment:
Okay so there are 2 things that we need to do here:

1. Only keep stuff in the try block that can raise. This function is currently too complex for what it's doing.
2. Instead of returning a list of softeners, transform it into a dict, with the `dsn` as key and the Softener as value. This way you're not bound to the index of a list. I think currently if you would add a third Softener, it could break